src/lem/circuit.rs

```diff
@@ -552,26 +646,194 @@ impl Func {
+                        let a = bound_allocations.get(a)?;
+                        let a_bits = a.hash().to_bits_le(&mut cs.namespace(|| "bitwise_and"))?;
+                        let mut trunc_bits = Vec::with_capacity(64);
+                        let mut b_rest = *b;
+                        for a_bit in a_bits {
+                            let b_bit = b_rest & 1;
+                            if b_bit == 1 {
+                                trunc_bits.push(a_bit.clone());
+                            } else {
+                                trunc_bits.push(Boolean::Constant(false))
+                            }
+                            b_rest /= 2;
+                        }
+                        let trunc = AllocatedNum::alloc(cs.namespace(|| "trunc"), || {
+                            let val = a
+                                .hash()
+                                .get_value()
+                                .map(|a| F::from_u64(a.to_u64_unchecked() & b))
+                                .unwrap();
+                            Ok(val)
+                        })?;
+                        enforce_pack(&mut cs.namespace(|| "enforce_trunc"), &trunc_bits, &trunc)?;
+                        let tag = g
+                            .global_allocator
+                            .get_or_alloc_const(cs, Tag::Expr(Num).to_field())?;
+                        let c = AllocatedPtr::from_parts(tag, trunc);
+                        bound_allocations.insert(tgt.clone(), c);
+                    }
+                    Op::DivRem64(tgt, a, b) => {
+                        let a = bound_allocations.get(a)?.hash();
+                        let b = bound_allocations.get(b)?.hash();
+                        let div_rem = a.get_value().and_then(|a| {
+                            b.get_value().map(|b| {
+                                if not_dummy.get_value().unwrap() {
+                                    let a = a.to_u64_unchecked();
+                                    let b = b.to_u64_unchecked();
+                                    (F::from_u64(a / b), F::from_u64(a % b))
+                                } else {
+                                    (F::ZERO, a)
+                                }
+                            })
+                        });
+                        let div =
+                            AllocatedNum::alloc(cs.namespace(|| "div"), || Ok(div_rem.unwrap().0))?;
+                        let rem =
+                            AllocatedNum::alloc(cs.namespace(|| "rem"), || Ok(div_rem.unwrap().1))?;
+
+                        let diff = sub(cs.namespace(|| "diff for slot {slot}"), b, &rem)?;
+                        implies_u64(cs.namespace(|| "div_u64"), not_dummy, &div)?;
+                        implies_u64(cs.namespace(|| "rem_u64"), not_dummy, &rem)?;
+                        implies_u64(cs.namespace(|| "diff_u64"), not_dummy, &diff)?;
+
+                        linear(cs, || "linear", b, &div, &rem, a);
```

Comment:
This usage shows why `linear` is not a great name for this function. I'd love a combination of name and argument name/order that makes everything more obvious, if possible. Can we come up with something better?

Reply:
renamed to  `enforce_product_and_sum`

Reply:
I don't see that rename